WireDomain/Sources/WireDomain/Event Processing/ConversationEventProcessor/ConversationTypingUsersTimeout.swift

```diff
@@ -0,0 +1,118 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireDataModel
+
+/// Keeping track of typing users timeouts
+final class ConversationTypingUsersTimeout: NSObject, ZMTimerClient {
+    
+    struct Key: Hashable {
+        let userObjectID: NSManagedObjectID
+        let conversationObjectID: NSManagedObjectID
+    }
+
+    static let defaultTimeout: TimeInterval = 60
+    
+    private var timeouts = [Key: Date]()
+    private var nextPruneDate: Date?
+    private var expirationTimer: ZMTimer?
+   
+    typealias VoidClosure = () async -> Void
+    var timerFiredCallback: VoidClosure?
+    
+    var firstTimeout: Date? {
+        timeouts.values.min()
+    }
+
+    func add(
+        _ user: NSManagedObjectID,
+        for conversation: NSManagedObjectID,
+        withTimeout timeout: Date
+    ) {
+        let key = Key(userObjectID: user, conversationObjectID: conversation)
+        timeouts[key] = timeout
+    }
+
+    func remove(
+        _ user: NSManagedObjectID,
+        for conversation: NSManagedObjectID
+    ) {
+        let key = Key(userObjectID: user, conversationObjectID: conversation)
+        timeouts.removeValue(forKey: key)
+    }
+
+    func contains(
+        _ user: NSManagedObjectID,
+        for conversation: NSManagedObjectID
+    ) -> Bool {
+        let key = Key(userObjectID: user, conversationObjectID: conversation)
+        return timeouts[key] != nil
+    }
+
+    func userIds(
+        in conversation: NSManagedObjectID
+    ) -> Set<NSManagedObjectID> {
+        let userIds = timeouts.keys
+            .filter { $0.conversationObjectID == conversation }
+            .map(\.userObjectID)
+
+        return Set(userIds)
+    }
+
+    func pruneConversationsThatHaveTimoutBefore(
+        date pruneDate: Date
+    ) -> Set<NSManagedObjectID> {
+        let keysToRemove = timeouts
+            .filter { $0.value < pruneDate }
+            .keys
+
+        keysToRemove.forEach {
+            timeouts.removeValue(forKey: $0)
+        }
+        
+        return Set(keysToRemove.map(\.conversationObjectID))
+    }
+    
+    /// Updates next prune date and fire a timer at that specific date.
+    /// When timer fires, the provided callback will be called - see `timerDidFire`
+
+    func updateExpirationIfNeeded() {
+        guard
+            let date = firstTimeout,
+            date != nextPruneDate
+        else {
+            return
+        }
+
+        expirationTimer?.cancel()
+        expirationTimer = ZMTimer(target: self)
+        expirationTimer?.fire(at: date)
+        nextPruneDate = date
+    }
+    
+    
+    func timerDidFire(_ timer: ZMTimer!) {
+        guard timer === expirationTimer else {
+            return
+        }
+        
+        Task { [timerFiredCallback] in
```

Comment:
do we need to hold on this task and cancel it if we're getting called again?

Reply:
if you mean holding a reference to that task and cancelling it when it is called again, I would say we don't need to to do it is a short lived task designed to be called every time the timer fires